File: frontend/de/unika/ipd/grgen/ir/ArrayVarAddItem.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 3.0
 * Copyright (C) 2003-2011 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ir;

public class ArrayVarAddItem extends EvalStatement {
	Variable target;
    Expression valueExpr;
    Expression indexExpr;

	public ArrayVarAddItem(Variable target, Expression valueExpr, Expression indexExpr) {
		super("array var add item");
		this.target = target;
		this.valueExpr = valueExpr;
		this.indexExpr = indexExpr;
	}

	public Variable getTarget() {
		return target;
	}

	public Expression getValueExpr() {
		return valueExpr;
	}

	public Expression getIndexExpr() {
		return indexExpr;
	}

	public void collectNeededEntities(NeededEntities needs)
	{
		needs.add(target);

		getValueExpr().collectNeededEntities(needs);

		if(getIndexExpr()!=null)
			getIndexExpr().collectNeededEntities(needs);

		if(getNext()!=null) {
			getNext().collectNeededEntities(needs);
		}
	}
}

File: frontend/de/unika/ipd/grgen/ir/stmt/array/ArrayVarAddItem.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ir.stmt.array
{
	using de.unika.ipd.grgen.ir;
	using Expression = de.unika.ipd.grgen.ir.expr.Expression;
	using Variable = de.unika.ipd.grgen.ir.pattern.Variable;
	using ContainerVarProcedureMethodInvocationBase = de.unika.ipd.grgen.ir.stmt.ContainerVarProcedureMethodInvocationBase;

	public class ArrayVarAddItem : ContainerVarProcedureMethodInvocationBase
	{
		internal Expression valueExpr;
		internal Expression indexExpr;

		public ArrayVarAddItem(Variable target, Expression valueExpr, Expression indexExpr)
			: base("array var add item", target)
		{
			this.valueExpr = valueExpr;
			this.indexExpr = indexExpr;
		}

		public virtual Expression ValueExpr
		{
			get
			{
				return valueExpr;
			}
		}

		public virtual Expression IndexExpr
		{
			get
			{
				return indexExpr;
			}
		}

		public override void CollectNeededEntities(NeededEntities needs)
		{
			base.CollectNeededEntities(needs);

			valueExpr.CollectNeededEntities(needs);

			if(indexExpr != null)
				indexExpr.CollectNeededEntities(needs);
		}
	}

}
